From babec3a50dd710d26b72f6c6d43bd79b04e954a6 Mon Sep 17 00:00:00 2001
From: Tanu Kaskinen <tanu.kaskinen@linux.intel.com>
Date: Tue, 28 Apr 2015 14:32:43 +0300
Subject: [PATCH] client-conf: Add allow-autospawn-for-root

Usually autospawning for root is a bad idea, since it can easily
interfere with other users' PulseAudio instances, but in embedded
environments where only root exists, autospawning is fine.

Upstream-Status: Denied [Upstream hasn't so far been convinced that
this feature is needed. The view is that running the graphical session
as root doesn't make sense even on embedded systems, or if it does,
running PulseAudio in the system mode should be a good solution for
that use case. If someone can prove otherwise, the patch would
probably be accepted.]

Signed-off-by: Tanu Kaskinen <tanu.kaskinen@linux.intel.com>
---
 man/pulse-client.conf.5.xml.in | 9 +++++++++
 src/pulse/client-conf.c        | 1 +
 src/pulse/client-conf.h        | 1 +
 src/pulse/client.conf.in       | 1 +
 src/pulse/context.c            | 2 +-
 5 files changed, 13 insertions(+), 1 deletion(-)

diff --git a/man/pulse-client.conf.5.xml.in b/man/pulse-client.conf.5.xml.in
index b88898c..e737c96 100644
--- a/man/pulse-client.conf.5.xml.in
+++ b/man/pulse-client.conf.5.xml.in
@@ -82,6 +82,15 @@ License along with PulseAudio; if not, see <http://www.gnu.org/licenses/>.
     </option>
 
     <option>
+      <p><opt>allow-autospawn-for-root=</opt> Allow autospawning also for root.
+      Takes a boolean value, defaults to <opt>no</opt>. If the <opt>autospawn
+      </opt> option is disabled, this option has no effect. Autospawning for
+      root is disabled by default, because running PulseAudio as root will
+      interfere with regular users' PulseAudio instances. This option should be
+      enabled only in environments where there are no regular users at all.</p>
+    </option>
+
+    <option>
       <p><opt>daemon-binary=</opt> Path to the PulseAudio daemon to
       run when autospawning. Defaults to a path configured at compile
       time.</p>
diff --git a/src/pulse/client-conf.c b/src/pulse/client-conf.c
index a3c9486..9f68ee5 100644
--- a/src/pulse/client-conf.c
+++ b/src/pulse/client-conf.c
@@ -138,6 +138,7 @@ void pa_client_conf_load(pa_client_conf *c, bool load_from_x11, bool load_from_e
         { "default-server",         pa_config_parse_string,   &c->default_server, NULL },
         { "default-dbus-server",    pa_config_parse_string,   &c->default_dbus_server, NULL },
         { "autospawn",              pa_config_parse_bool,     &c->autospawn, NULL },
+        { "allow-autospawn-for-root", pa_config_parse_bool,   &c->allow_autospawn_for_root, NULL },
         { "cookie-file",            pa_config_parse_string,   &c->cookie_file_from_client_conf, NULL },
         { "disable-shm",            pa_config_parse_bool,     &c->disable_shm, NULL },
         { "enable-shm",             pa_config_parse_not_bool, &c->disable_shm, NULL },
diff --git a/src/pulse/client-conf.h b/src/pulse/client-conf.h
index 7691ec7..19db7ed 100644
--- a/src/pulse/client-conf.h
+++ b/src/pulse/client-conf.h
@@ -38,6 +38,7 @@ typedef struct pa_client_conf {
     char *cookie_file_from_application;
     char *cookie_file_from_client_conf;
     bool autospawn, disable_shm, disable_memfd, auto_connect_localhost, auto_connect_display;
+    bool allow_autospawn_for_root;
     size_t shm_size;
 } pa_client_conf;
 
diff --git a/src/pulse/client.conf.in b/src/pulse/client.conf.in
index 26b7790..69830ef 100644
--- a/src/pulse/client.conf.in
+++ b/src/pulse/client.conf.in
@@ -23,6 +23,7 @@
 ; default-dbus-server =
 
 ; autospawn = yes
+; allow-autospawn-for-root = no
 ; daemon-binary = @PA_BINARY@
 ; extra-arguments = --log-target=syslog
 
diff --git a/src/pulse/context.c b/src/pulse/context.c
index 69be5f4..d6e13e8 100644
--- a/src/pulse/context.c
+++ b/src/pulse/context.c
@@ -1027,7 +1027,7 @@ int pa_context_connect(
     if (!(flags & PA_CONTEXT_NOAUTOSPAWN) && c->conf->autospawn) {
 
 #ifdef HAVE_GETUID
-        if (getuid() == 0)
+        if (!c->conf->allow_autospawn_for_root && getuid() == 0)
             pa_log_debug("Not doing autospawn since we are root.");
         else {
             c->do_autospawn = true;
-- 
2.8.1

